identity-service/src/solana-client.js

```diff
@@ -0,0 +1,172 @@
+const config = require('./config')
+const solanaWeb3 = require('@solana/web3.js')
+const keccak256 = require('keccak256')
+const secp256k1 = require('secp256k1')
+const borsh = require('borsh')
+
+const VALID_SIGNER = config.get('solanaValidSigner')
+const AUDIUS_ETH_REGISTRY_PROGRAM = config.get('solanaAudiusEthRegistryAddress') ? new solanaWeb3.PublicKey(
+  config.get('solanaAudiusEthRegistryAddress')
+) : null
+const TRACK_LISTEN_PROGRAM = config.get('solanaTrackListenCountAddress') ? new solanaWeb3.PublicKey(
+  config.get('solanaTrackListenCountAddress')
+) : null
+const INSTRUCTIONS_PROGRAM = new solanaWeb3.PublicKey(
+  'Sysvar1nstructions1111111111111111111111111'
+)
+const CLOCK_PROGRAM = new solanaWeb3.PublicKey(
+  'SysvarC1ock11111111111111111111111111111111'
+)
+
+class Assignable {
+  constructor (data) {
+    Object.assign(this, data)
+  }
+}
+
+class TrackData extends Assignable {}
+class InstructionArgs extends Assignable {}
+class InstructionEnum extends Assignable {}
+
+const trackDataSchema = new Map([
+  [
+    TrackData,
+    {
+      kind: 'struct',
+      fields: [
+        ['user_id', 'string'],
+        ['track_id', 'string'],
+        ['source', 'string'],
+        ['timestamp', 'u64']
+      ]
+    }
+  ]
+])
+
+const instructionSchema = new Map([
+  [
+    InstructionEnum,
+    {
+      kind: 'enum',
+      field: 'choose',
+      values: [['instruction', InstructionArgs]]
+    }
+  ],
+  [
+    InstructionArgs,
+    {
+      kind: 'struct',
+      fields: [
+        ['track_data', TrackData],
+        ['signature', [64]],
+        ['recovery_id', 'u8']
+      ]
+    }
+  ],
+  [
+    TrackData,
+    {
+      kind: 'struct',
+      fields: [
+        ['user_id', 'string'],
+        ['track_id', 'string'],
+        ['source', 'string'],
+        ['timestamp', 'u64']
+      ]
+    }
+  ]
+])
+
+let solanaConnection = new solanaWeb3.Connection(config.get('solanaEndpoint'))
+let feePayer
+
+function getFeePayer () {
+  if (!feePayer) {
+    feePayer = config.get('solanaFeePayerWallet') ? new solanaWeb3.Account(config.get('solanaFeePayerWallet')) : null
+  }
+  return feePayer
+}
+
+async function createAndVerifyMessage (
+  validSigner,
+  privateKey,
+  userId,
+  trackId,
+  source
+) {
+  validSigner = validSigner || VALID_SIGNER
+
+  let privKey = Buffer.from(privateKey, 'hex')
+  let pubKey = secp256k1.publicKeyCreate(privKey, false).slice(1)
+
+  let validSignerPubK = new solanaWeb3.PublicKey(validSigner)
+  let accInfo = await solanaConnection.getAccountInfo(validSignerPubK)
+  let signerGroup = new solanaWeb3.PublicKey(
+    accInfo.data.toJSON().data.slice(1, 33)
+  ) // cut off version and eth address from valid signer data
+
+  let trackData = new TrackData({
```

Comment:
is this the standard way of serializing to solana? just using an arbitrary assignable object? might be nice if we had some more structure around what TrackData is so we don't accidentally construct it with something else?

Reply:
I added a constructor to make this somewhat clearer